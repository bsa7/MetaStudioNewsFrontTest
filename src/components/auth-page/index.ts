import AuthPage from './auth.container'

export { AuthPage }
