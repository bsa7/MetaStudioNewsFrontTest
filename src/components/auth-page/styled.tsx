import styled from 'styled-components';

export const Root = styled.div`
  display: flex;
  align-items: center;
  position: relative;
  padding: 40px;
`
export const Row = styled.div`
  display: flex;
  flex-direction: row;
`

export const Column = styled.div`
  display: flex;
  flex-direction: column;
`
