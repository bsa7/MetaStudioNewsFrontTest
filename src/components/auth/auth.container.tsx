import * as React from 'react'
import { IAuthProps, IAuthState, Auth } from './auth.view'

export class AuthContainer extends React.Component<IAuthProps, IAuthState> {
  state: IAuthState = {
  }

  render() {
    return (
      <Auth />
    )
  }
}
