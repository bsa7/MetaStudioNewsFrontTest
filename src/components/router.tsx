import * as React from 'react'
import { pathSettings } from '@config/routes'
import { EnumMap, HashMap, LocationInfo, PathSetting } from '@lib/common-defs'
import * as ApplicationPages from '@components/index'
import { currentLocation } from '@lib/isomorphic-helper'
import { getPropInSafe, traceError } from '@lib/sys-helper'
import { ApplicationPageName } from '@lib/common-defs'
import { Route } from 'react-router-dom'

type IRouterProps = {
}

type IRouterState = {}

type RouteSetting = {
  hostname: string;
  locationRegexp: string;
  pathname: string;
  pathSetting: PathSetting;
}

export class Router extends React.Component<IRouterProps, IRouterState> {
  private routes: Array<RouteSetting>
  private pathSettings: HashMap<PathSetting>
  public status: number
  public links: EnumMap<typeof pathSettings, string>
  static initialized: boolean
  static instance: Router

  constructor(props?: IRouterProps) {
    super(props)
    if (Router.initialized) {
      return Router.instance
    } else {
      this.init()
      Router.initialized = true
      Router.instance = this
    }
  }

  private init(): void {
    this.pathSettings = pathSettings
    this.routes = []
    Object.keys(this.pathSettings).forEach((pathSettingKey) => {
      const pathSetting = this.pathSettings[pathSettingKey]
      let locationRegexp = `^${pathSetting.pathname}$`
      this.pathSettings[pathSettingKey].key = pathSettingKey
      this.pathSettings[pathSettingKey].hostname = (pathSetting.params || {}).domain
      this.pathSettings[pathSettingKey].locationRegexp = locationRegexp
      this.pathSettings[pathSettingKey].pathname = pathSetting.pathname
      this.routes.push({
        hostname: (pathSetting.params || {}).domain,
        locationRegexp,
        pathname: pathSetting.pathname,
        pathSetting,
      })
    })
    this.links = {}
    Object.keys(this.pathSettings).forEach((key) => this.links[key] = key)
  }

  private findRoute = (locationInfo: LocationInfo = currentLocation.locationInfo()): RouteSetting => {
    let route = this.routes.find((route: RouteSetting) => {
      const checkRegexp = new RegExp(route.locationRegexp)
      const pathnameMatched = checkRegexp.test(locationInfo.pathname)
      const hostMatched = (route.hostname && route.hostname == locationInfo.hostname) || !route.hostname
      return pathnameMatched && hostMatched
    })
    if (typeof route === 'undefined') {
      route = this.routes.slice(-1)[0]
      traceError({ message: `this.findRoute error: route ${locationInfo.pathname} not found.` })
    }
    this.status = getPropInSafe(route, (r) => r.pathSetting.params.status, 200)
    return route
  }

  public get currentPathSetting() {
    const currentRoute = this.findRoute()
    return currentRoute.pathSetting
  }

  render() {
    if (!Router.initialized) this.init()
    const currentRoute: RouteSetting = this.findRoute()
    const currentRoutePageName: string = currentRoute.pathSetting.componentName
    const CurrentPage = ApplicationPages[currentRoutePageName as ApplicationPageName]

    return <CurrentPage />
  }
}

export const router = new Router()
