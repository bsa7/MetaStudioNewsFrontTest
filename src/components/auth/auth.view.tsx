import * as React from 'react'
import { Root } from './styled'

export interface IAuthProps {
}

export interface IAuthState {
}

export class Auth extends React.Component<IAuthProps, IAuthState> {
  state: IAuthState = {
  }

  render() {
    return (
      <Root>
        <h1>Auth, world!</h1>
      </Root>
    )
  }
}
