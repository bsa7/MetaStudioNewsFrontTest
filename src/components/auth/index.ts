import { AuthContainer } from './auth.container'

export const Auth = AuthContainer
