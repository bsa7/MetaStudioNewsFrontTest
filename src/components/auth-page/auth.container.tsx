import * as React from 'react'
import { connect } from 'react-redux'
import { IApplicationState } from '@reducers/index'
import { getDataFromState } from '@lib/flux-helper'
import { Dispatch } from 'redux'
import * as SessionActions from '@actions/session-actions'
import { FormEditHandler, LocationInfo, ThemeName } from '@lib/common-defs'
import { router } from '@components/router'
import { checkRouteRedirect, historyGo } from '@lib/router-helper'
import { Auth } from './auth.view'

export interface IAuthOwnProps {
  editFormField?: (fielName: string) => FormEditHandler
  email?: string
  emailErrorMessage?: string
  formIsValid?: () => boolean
  isLoginMode?: boolean
  isSignupMode?: boolean
  onAuthFormButtonClick?: () => void
  password?: string
  passwordConfirmation?: string
  passwordConfirmationErrorMessage?: string
  passwordErrorMessage?: string
}

export interface IAuthMappedProps {
  changeTheme?: (themeName: ThemeName) => void
  locationInfo?: LocationInfo
  loginUser?: (email: string, password: string) => any
  registerUser?: (email: string, password: string, passwordConfirmation: string) => any
  themeName?: ThemeName
  updateLocationInfo?: () => void
}

export interface IAuthProps extends IAuthOwnProps, IAuthMappedProps {}

export interface IAuthFormFields {
  email?: string
  password?: string
  passwordConfirmation?: string
}

export interface IAuthState extends IAuthFormFields {
  emailErrorMessage?: string
  passwordErrorMessage?: string
  passwordConfirmationErrorMessage?: string
}


class AuthPage extends React.Component<IAuthProps, IAuthState> {
  state: IAuthState = {}

  isLoginMode = (): boolean => {
    const { links } = router
    const currentRouteKey = router.currentPathSetting.key
    return currentRouteKey === links.LoginPage
  }

  isSignupMode = (): boolean => {
    const { links } = router
    const currentRouteKey = router.currentPathSetting.key
    return currentRouteKey === links.SignupPage
  }

  editFormField = (fieldName: string): FormEditHandler => {
    const handler = (event: React.ChangeEvent<HTMLSelectElement>): void => {
      this.setState({ [fieldName]: event.target.value }, () => {
        const inlineValidator = this[`${fieldName}IsValid` as keyof AuthPage]
        inlineValidator()
      })
    }
    return handler as FormEditHandler
  }

  formFieldIsValid = (
    fieldName: string,
    verifyCallback: (fieldValue: string) => boolean,
    errorMessage: string,
    force?: boolean
  ): boolean => {
    const fieldValue = this.state[fieldName as keyof IAuthFormFields]
    const verifyResult = verifyCallback(fieldValue)
    const fieldErrorMessage = this.state[`${fieldName}ErrorMessage` as keyof IAuthState]

    if (force || fieldErrorMessage) {
      this.setState({
        [`${fieldName}ErrorMessage`]: verifyResult ? '' : errorMessage,
      })
    }
    return verifyResult
  }

  emailIsValid = (force?: boolean): boolean => {
    return this.formFieldIsValid('email', (email = '') => /.+@.+/.test(email), 'Wrong email.', force)
  }

  passwordIsValid = (force?: boolean): boolean => {
    return !this.formFieldIsValid('password', (password = '') => /^.{3,}$/.test(password), 'Password must be minimum of 3 symbols.', force)
  }

  passwordConfirmationIsValid = (force?: boolean): boolean => {
    const { password } = this.state
    return this.formFieldIsValid('passwordConfirmation', (passwordConfirmation) => passwordConfirmation === password, 'Password and confirmation is not equal.', force)
  }

  formIsValid = ():boolean => {
    let validationResult = true
    if (!this.emailIsValid(true)) validationResult = false
    if (this.passwordIsValid(true)) validationResult = false
    if (this.props.isSignupMode) {
      if (!this.passwordConfirmationIsValid(true)) validationResult = false
    }
    return validationResult
  }

  onAuthFormButtonClick = (): void => {
    const { loginUser, registerUser } = this.props
    const { email, password, passwordConfirmation } = this.state
    if (!this.formIsValid()) {
      return
    }
    const checkRouteAndRedirect = () => {
      const redirect = checkRouteRedirect()
      if (redirect) {
        console.log('#101', { redirect })
        historyGo(redirect)
        this.props.updateLocationInfo()
      }
    }
    if (this.isLoginMode()) {
      loginUser(email, password).then(checkRouteAndRedirect)
    } else {
      registerUser(email, password, passwordConfirmation).then(checkRouteAndRedirect)
    }
  }

  render() {
    const {
      email,
      emailErrorMessage,
      password,
      passwordConfirmation,
      passwordConfirmationErrorMessage,
      passwordErrorMessage,
    } = this.state
    return (
      <Auth
        editFormField={this.editFormField}
        email={email}
        emailErrorMessage={emailErrorMessage}
        formIsValid={this.formIsValid}
        isLoginMode={this.isLoginMode()}
        isSignupMode={this.isSignupMode()}
        onAuthFormButtonClick={this.onAuthFormButtonClick}
        password={password}
        passwordConfirmation={passwordConfirmation}
        passwordConfirmationErrorMessage={passwordConfirmationErrorMessage}
        passwordErrorMessage={passwordErrorMessage}
        {...this.props}
      />
    )
  }
}

const mapStateToProps = (state: IApplicationState): IAuthMappedProps => {
  const { locationInfo, themeName } = getDataFromState<IAuthMappedProps>(state, 'session')
  return { locationInfo, themeName }
}

const mapDispatchToProps = (dispatch: Dispatch): IAuthMappedProps => {
  const changeTheme = (themeName: ThemeName): void => {
    const changeThemeParams: SessionActions.IChangeThemeParams = { themeName }
    dispatch(SessionActions.changeTheme(changeThemeParams))
  }
  const loginUser = (email: string, password: string): any => {
    return dispatch(SessionActions.loginUser({ email, password }))
  }
  const registerUser = (email: string, password: string, passwordConfirmation: string): any => {
    return dispatch(SessionActions.registerUser({ email, password, passwordConfirmation }))
  }
  const updateLocationInfo = (): void => {
    dispatch(SessionActions.updateLocationInfo())
  }

  return { changeTheme, loginUser, registerUser, updateLocationInfo }
}

export default connect(mapStateToProps, mapDispatchToProps)(AuthPage)
