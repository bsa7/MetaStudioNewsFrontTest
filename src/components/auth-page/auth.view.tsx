import * as React from 'react'
import { Link } from 'react-router-dom'
import { router } from '@components/router'
import { seoLink } from '@lib/router-helper'
import { Layout } from '@components/layout'
import { Column, Root, Row } from './styled'

export interface IAuthProps {
}

export interface IAuthState {
}

export class Auth extends React.Component<IAuthProps, IAuthState> {
  state: IAuthState = {}

  render() {
    const currentRouteKey = router.currentPathSetting.key
    const titleText: string = `${currentRouteKey === router.links.LoginPage ? 'Login' : 'Signup'}, my Friend!!!`
    const linkText: string = currentRouteKey === router.links.LoginPage
      ? 'Not registered yet? Signup!'
      : 'Already registered? Login!'

    return (
      <Layout>
        <Root>
          <Column>
            <Row>
            <h1>{titleText}</h1>
            </Row>
            <Row>
              <Link
                to={seoLink(currentRouteKey === router.links.LoginPage ? router.links.SignupPage : router.links.LoginPage)}
              >
                {linkText}
              </Link>
            </Row>
          </Column>
        </Root>
      </Layout>
    )
  }
}
