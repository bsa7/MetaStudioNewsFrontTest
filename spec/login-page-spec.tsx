import React from 'react'
import Enzyme, { shallow, ShallowWrapper } from 'enzyme'
import { create, ReactTestRenderer } from 'react-test-renderer'
import { AuthPage } from '@components/auth-page/index'
import { IAuthProps } from '@components/auth-page/auth.view'
import { router } from '@components/router'
import toJson from 'enzyme-to-json'
import { enzymeConfigureAdapter } from './setup-specs'

enzymeConfigureAdapter()
let wrapper: ShallowWrapper<IAuthProps>
let snapshot: ReactTestRenderer

// beforeEach(() => {
//   const component = <AuthPage />

//   wrapper = shallow(component)
//   snapshot = create(component)
// })

describe('<AuthPage />', () => {
  it('renders as expected', () => {
    const component = <AuthPage />

    wrapper = shallow(component)
    snapshot = create(component)

    expect(toJson(wrapper)).toMatchSnapshot()
  });
})