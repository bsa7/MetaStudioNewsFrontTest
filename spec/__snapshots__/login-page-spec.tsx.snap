// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AuthPage /> renders as expected 1`] = `undefined`;
